kong/tools/queue.lua

```diff
@@ -0,0 +1,408 @@
+--     return true
+--   end
+--
+--   local handler_conf = {...}  -- configuration for queue handler
+--   local queue_conf =          -- configuration for the queue itself (defaults shown unless noted)
+--     {
+--       name = "example",       -- name of the queue (required)
+--       batch_max_size = 10,    -- maximum number of entries in one batch (default 1)
+--       max_delay = 1,          -- maximum number of seconds after first entry before a batch is sent
+--       capacity = 10,          -- maximum number of entries on the queue (default 10000)
+--       string_capacity = 100,  -- maximum number of bytes on the queue (default nil)
+--       max_retry_time = 60,    -- maximum number of seconds before a failed batch is dropped
+--       max_retry_delay = 60,   -- maximum delay between send attempts
+--     }
+--
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Some value")
+--   Queue.enqueue(queue_conf, handler, handler_conf, "Another value")
+--
+-- Given the example above,
+--
+-- * If the two `enqueue()` invocations are done within `max_delay` seconds, they will be passed to the
+--   handler function together in one batch.  The maximum number of entries in one batch is defined
+--   by the `batch_max_size` parameter.
+-- * The `capacity` parameter defines how many entries can be waiting on the queue for transmission.  If
+--   it is exceeded, the oldest entries on the queue will be discarded when new entries are queued.  Error
+--   messages describing how many entries were lost will be logged.
+-- * The `string_capacity` parameter, if set, indicates that no more than that number of bytes can be
+--   waiting on a queue for transmission.  If it is set, only string entries must be queued and an error
+--   message will be logged if an attempt is made to queue a non-string entry.
+-- * When the `handler` function does not return a true value for a batch, it is retried for up to
+--   `max_retry_time` seconds before the batch is deleted and an error is logged.  Retries are organized
+--   by the queue library using an exponential back-off algorithm with the maximum time between retries
+--   set to `max_retry_delay` seconds.
+
+local workspaces = require "kong.workspaces"
+local semaphore = require "ngx.semaphore"
+
+local DEBUG = ngx.DEBUG
+local INFO = ngx.INFO
+local WARN = ngx.WARN
+local ERR = ngx.ERR
+
+
+local function now()
+  ngx.update_time()
+  return ngx.now()
+end
+
+
+local Queue = {
+  CAPACITY_WARNING_THRESHOLD = 0.8, -- Threshold to warn that the queue capacity limit is reached
+}
+
+
+local Queue_mt = {
+  __index = Queue
+}
+
+
+local function make_queue_key(name)
+  return string.format("%s.%s", workspaces.get_workspace_id(), name)
+end
+
+
+local queues = setmetatable({}, {
+  __newindex = function (self, name, queue)
+    return rawset(self, make_queue_key(name), queue)
+  end,
+  __index = function (self, name)
+    return rawget(self, make_queue_key(name))
+  end
+})
```

Comment:
We cannot say at what point we run out of resources for queues.  Neither is it possible to say how many queues are too many, particularly because they only exist while they're in use.  You can have 10 million queues that are not used, or maybe a couple of thousand if they're all in use at the same time.

For capacity planning, it is necessary to experimentally find the limits of the system - That is true for all parts of Kong, not just for queues.

Reply:
We cannot release such a feature without some numbers that tell the user what sort of memory usage to expect.

Please work with @kikito and SDETs to figure out the benchmark numbers we need and then use SDET and other automation to come up with those. I'm not looking for accurate data but a trend and rough sizing estimation in the order of 100s of MB (10s of MB would be even better).

Reply:
It is difficult for us to find a suitable figure for this limitation.

If it was too small it could result in customers not being able to run their services, if it was too large it didn't make sense. Deciding on this number required a lot of data to back it up. 

And once we decided s to implement this limit, we had to provide a configuration to adjust the number, which I think introduced additional unnecessary complexity.

Reply:
> We cannot release such a feature without some numbers that tell the user what sort of memory usage to expect.

@hbagdi Users don't interact with queues, they interact with plugin instances.  With the new parameters, they have a way to configure queue sharing, which reduces memory usage.  For their capacity planning purposes, the number of plugins is the first thing that they need to limit.  In what document do we document the memory requirements for each of the plugins?  That would be the document in which we mention that the per-plugin memory usage can be reduced by queue sharing.

Reply:
@hbagdi 

> Is there a limit on the number of queues? Should there be one?

There's an upper limit - we can never have more queues than plugin instances. Because each plugin instance can reference, at most, one queue (and most don't reference any at all).

I suspect the overhead of using queues will be negligible compared to the memory needed to hold the batched items that need to be processed. See below.

@hanshuebner perhaps we can give an "equation" based on some assumptions on the "queue item size" and the expected traffic for each plugin.

For http_log, taking [this log format](https://docs.konghq.com/hub/kong-inc/http-log/#log-format) example as a base, we need:
* ~16 Lua tables. Assuming each table "costs" 40 bytes (pulled that out of the internet, probably is a good enough ceiling), that's ~640 bytes
* If we delete the spaces, quotes, {, }, [ and ] we get 1383 bytes. That's a good higher ceiling for the "contents" of the table, which are mostly Lua strings. If we assume that half of them will be interned that's another 690 bytes.
* We also need to allocate a string for the json representation of the item, which is ~1788 bytes (the string with only spaces removed).

That gives us 640 + 690 + 1788 = 3118, or ~3KB per batch item.

Then we can put the memory usage in terms of traffic: "The HTTP Log plugin will allocate ~3KB of memory per request while it is processing each request. If your server's traffic is 10000 requests per second, it will increase memory usage by ~30MB every second, until it sends that information (in batches) to its backend. The backed needs to support that throughput".

As you can see, we don't care about how many queues there are. We care about how much each request "costs", and how much traffic there is.

It should be relatively straightforward to roughly confirm the queue item size with a test that makes the batch size 10000, makes 9999 requests, and then compares memory usage, before and after.

Reply:
I don't think it makes sense to talk about concrete bytes of memory usage, given that we have no control over how garbage memory accumulates in Kong's heap and is never freed.  The dynamic behavior of Kong depends too much on the actual load pattern to be able to come up with these kinds of numbers.  Also, as the http-log pattern logs whatever has been put into the log context by other plugins, calculating capacity planning numbers based on examples and rough estimates from the Internet are bound to result in incorrect numbers.

Given that we cannot directly determine Kong's run-time memory usage purely based on the configuration and other external parameters, we should rather invest in creating a document that describes a methodology how to perform capacity planning based on an actual configuration and load generators.  That capacity planning methodology should include information on how to plan in plugins that use queues, and highlight that measurements need to be taken with queue consumers being unresponsive.

To my knowledge, no capacity planning document exists that could be augmented this way, our old queue implementation does not have any memory limits and we have no internal metrics that'd allow us to provide meaningful parameter based tuning information.  Creating a good capacity planning document seems to be a separate task, and require a larger scope than the queue reimplementation.

Reply:
As discussed, queues will be mentioned in our [sizing document](https://docs.konghq.com/gateway/latest/production/sizing-guidelines/), but no detailed byte-level numbers will be provided.